config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncapsulatedAssertionBuilder.java

```diff
@@ -0,0 +1,178 @@
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static com.ca.apim.gateway.cagatewayconfig.util.policy.PolicyXMLElements.*;
+import static com.ca.apim.gateway.cagatewayconfig.util.xml.DocumentUtils.createElementWithAttribute;
+import static com.ca.apim.gateway.cagatewayconfig.util.xml.DocumentUtils.getSingleChildElement;
+
+public class EncapsulatedAssertionBuilder implements PolicyAssertionBuilder {
+    private static final Logger LOGGER = Logger.getLogger(EncapsulatedAssertionBuilder.class.getName());
+    static final String ENCASS_NAME = "encassName";
+    static final String STRING_VALUE = "stringValue";
+    static final String BOOLEAN_VALUE = "booleanValue";
+    public static final String ZERO_GUID = "00000000-0000-0000-0000-000000000000";
+
+    @Override
+    public void buildAssertionElement(Element encapsulatedAssertionElement, PolicyBuilderContext policyBuilderContext) throws DocumentParseException {
+        final Bundle bundle = policyBuilderContext.getBundle();
+        final AnnotatedBundle annotatedBundle = policyBuilderContext.getAnnotatedBundle();
+        final Policy policy = policyBuilderContext.getPolicy();
+        if (encapsulatedAssertionElement.hasAttribute(ENCASS_NAME)) {
+            final String encassName = encapsulatedAssertionElement.getAttribute(ENCASS_NAME);
+            Encass encass = getEncass(bundle, encassName, annotatedBundle);
+            final String guid = findEncassReferencedGuid(policy, encass, encapsulatedAssertionElement, encassName);
+            updateEncapsulatedAssertion(policyBuilderContext, encapsulatedAssertionElement, encass, encassName, guid);
+        } else if (!isNoOpIfConfigMissing(encapsulatedAssertionElement)) {
+            Element guidElement = getSingleChildElement(encapsulatedAssertionElement, ENCAPSULATED_ASSERTION_CONFIG_GUID, true);
+            Element nameElement = getSingleChildElement(encapsulatedAssertionElement, ENCAPSULATED_ASSERTION_CONFIG_NAME, true);
+            throw new EntityBuilderException("No encassName specified for encass in policy: '" + policy.getPath() + "' GUID: '" + (guidElement != null ? guidElement.getAttribute(STRING_VALUE) : null) + "' Name: '" + (nameElement != null ? nameElement.getAttribute(STRING_VALUE) : null) + "'");
+        } else {
+            LOGGER.log(Level.FINE, "No encassName specified for encass in policy: \"{0}\". Since NoOp is true, this will be treated as a No Op.", policy.getPath());
+        }
+    }
+
+    private Encass getEncass(Bundle bundle, String name, AnnotatedBundle annotatedBundle) {
+        LOGGER.log(Level.FINE, "Looking for referenced encass: {0}", name);
+        final AtomicReference<Encass> referenceEncass = new AtomicReference<>(bundle.getEncasses().get(name));
```

Comment:
Just curious, what's the use of `AtomicReference` here?

Reply:
non final variables cannot be used in lambda expression so wrapping it with atomicaReference to set the value in lambda